src/Data/ElfEdit/Prim/HeaderInfo.hs

```diff
@@ -251,37 +252,67 @@ data Symtab w =
 
 deriving instance Show (ElfWordType w) => Show (Symtab w)
 
--- | Decodes the static symbol table using elf header info
+-- | Decodes the static symbol table using elf header info.
 --
 -- If no symbol table is present then @Nothing@ is returned.
 decodeHeaderSymtab :: ElfHeaderInfo w -> Maybe (Either SymtabError (Symtab w))
 decodeHeaderSymtab elf = elfClassInstances (headerClass (header elf)) $ do
+  let shdrs = headerShdrs elf
+  let symtabs = V.filter (\s -> shdrType s == SHT_SYMTAB) shdrs
+  when (V.length symtabs == 0) $ Nothing
+  Just $ decodeSymbolTable elf symtabs
+
+-- | Decodes the dynamic symbol table using ELF header info.
+--
+-- If no dynamic symbol table is present then @Nothing@ is returned.
+--
+-- The functionality of 'decodeHeaderDynsym' largely overlaps with what the
+-- @dynamicEntries@ and @dynSymEntry@ functions provide, but with some minor
+-- differences:
+--
+-- * The API for 'decodeHeaderDynsym' is much more direct, as it only requires
+--   an 'ElfHeaderInfo'.
+--
+-- * Unlike @dynSymEntry@, 'decodeHeaderDynsym' does not compute symbol version
+--   information.
+decodeHeaderDynsym :: ElfHeaderInfo w -> Maybe (Either SymtabError (Symtab w))
+decodeHeaderDynsym elf = elfClassInstances (headerClass (header elf)) $ do
+  let shdrs = headerShdrs elf
+  let dynSymtabs = V.filter (\s -> shdrType s == SHT_DYNSYM) shdrs
+  when (V.length dynSymtabs == 0) $ Nothing
```

Comment:
Nitpick: Unnecessary `$` operator

Reply:
Ah, this is a leftover from how `decodeHeaderSymtab` is was implemented prior to this patch. I'll remove both of these unnecessary `($)`s in a separate commit.